File: Leetcode/October_LeetCoding_Challenge/Day15_Rotate_Array/Day15_Rotate_Array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        int n = nums.size();
         k = k % n;  
        reverse(nums.begin(), nums.end()); 
        reverse(nums.begin(), nums.begin()+ k);
        reverse(nums.begin() +  k, nums.end());

}
};

// Array : 1 2 3 4 5 , k = 2, Output : 4 5 1 2 3
// Reverse array : 5 4 3 2 1

// Reverse 5 4 --> 4 5
// Now Reverse 1 2 3 --> 3 2 1

// Final Answer - 4 5 1 2 3

File: Leetcode/October_LeetCoding_Challenge/Day15_Rotate_Array/Day15_Rotate_Array.java
class Solution {
    public void rotate(int[] nums, int k) {
            k %= nums.length;
            reverse(nums, 0, nums.length - 1);
            reverse(nums, 0, k- 1);
            reverse(nums, k, nums.length - 1);
        
    }
    void reverse(int[] a, int start, int end){
        
        while(start < end){
            int temp = a[start];
            a[start] = a[end];
            a[end] = temp;
            start++;end--;
        }
    }
}

